### DISCUSSION TITLE
Is there a way to query a node that is the child of some root node?

### DISCUSSION THREAD

**Discussion by SalBakraa** [nvim-treesitter/nvim-treesitter#2166] [Q&A] [answered]
For example in the language I am querying, a variable is only builtin if it is inside of a lambda.

```kotlin
"Sample".map { it } // Here `it` is builtin variable
val it = 5 // Here `it` is a regular variable
```

for a construct like this

``` kotlin
val it = { it + (it * (it / it)) }
```

with this parse tree

```
property_declaration [0, 0] - [0, 33]
	variable_declaration [0, 4] - [0, 5]
		simple_identifier [0, 5] - [0, 5]
	lambda_literal [0, 8] - [0, 33]
		statements [0, 10] - [0, 31]
			additive_expression [0, 10] - [0, 31]
				simple_identifier [0, 10] - [0, 12]
				parenthesized_expression [0, 15] - [0, 31]
					multiplicative_expression [0, 16] - [0, 30]
						simple_identifier [0, 16] - [0, 18]
						parenthesized_expression [0, 21] - [0, 30]
							multiplicative_expression [0, 22] - [0, 29]
								simple_identifier [0, 22] - [0, 24]
								simple_identifier [0, 27] - [0, 29]
```
How would I match all simple_identifiers under the lambda_literal?

**Comment by theHamsta** [comment, marked as answer]
No, there is no built-in way to query this. This plugin has a predicate called `#has-ancester?` that does what you want but it has to use `:parent()` iteratively on nodes. `:parent()` has become inefficient after tree-sitter removed the parent cache.

**Reply by SalBakraa** [reply]
Are there any plans for the feature to be re-implemented? Because kotlin has multiple cases where identifiers hold a special meaning depending on their scope.

I guess I'll leave the option commented out for users that don't mind the performance hit.

Also thank you.

**Reply by theHamsta** [reply]
This is probably best implemented in tree-sitter core and it also has been requested there various times. Maybe we should also just walk the tree once to create a cache for everyone's parent which would also be useful for highlighting of definitions and references.

**Reply by SalBakraa** [reply]
Is it possible to create the cache in a query file? I haven't really seen any examples of someone trying to do that.

**Reply by theHamsta** [reply]
No, that has to be done in Lua code. Predicates and directives can execute arbitrary Lua functions once registered https://neovim.io/doc/user/treesitter.html#lua-treesitter-predicates but it would probably make sense to implement this in C directly in the tree-sitter. If you want to you can inspect the Lua source of `#has-ancestor?` or to collect this cache once required for the first time in some incremental manner.

You can try to measure whether using `#has-ancestor` has an impact on query performance at all. But the query is problematic as it would run a potential costly operation on almost all nodes

**Reply by SalBakraa** [reply]
Oh, I didn't know that I had the ability to add my custom predicates, when I read the query syntax documentation I thought predicates had to be hard coded by the tree-sitter engine.

Also thank you so much for the documentation link, I spent so much time trying to find the list of predicates in this repository that I just gave up.

I think the best to resolve such queries while keep costs down is to have the query handler be context/scope aware, and having queries be run only if defined scope or context is entered. I am not sure if such a feature is even feasible.

**Reply by theHamsta** [reply]
We have more than the ones listed in core. Just look into https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/query_predicates.lua

**Reply by SalBakraa** [reply]
From what I can tell, `has-parent` & `has-ancestor` are quite straight forward, `nth` checks if this node is the nth child of it's parent and it is 0-indexed. `is` and `has-type` I don't really get since I am not really familiar with lua.

Also it would do well if such predicates were mentioned in the [CONTRIBUTING.md](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md) instead of needing to dig through the source for them.

**Reply by theHamsta** [reply]
We've usptreamed all the predicates that we considered essential. We eventually should remove all other predicates. Nth works the same as in scheme/commonlisp and was only important when there was first?/last?. Now tree-sitter has anchors which made those functions unnecessary `.`

I can make a PR that removes that module and Al usages in dependent plugins.

`is?` is not documented because we had performance problems (there is PR that mitigates that problem). It works the same as in Atom. It is able to ask for the type of the definition of this node (whether the definition is a function or a parameter or a function). This enables file-local semantic highlighting.
